core/src/main/java/com/yahoo/imapnio/async/response/ImapResponseMapper.java

```diff
@@ -442,18 +477,116 @@ private SearchResult parseToSearchResult(@Nonnull final IMAPResponse[] ir) throw
+            Long highestModSeq = null;
+
+            for (final IMAPResponse sr: ir) {
+                sr.skipSpaces();
+                if (sr.keyEquals("FETCH")) {
+                    fetchResponses.add(parseToFetchResponse(sr, extensionItems));
+                } else if (sr.readByte() == (byte) L_BRACKET) { // HIGHESTMODSEQ or MODIFIED
+                    final String responseCode = sr.readAtom();
+                    if (sr.isOK() && responseCode.equalsIgnoreCase("HIGHESTMODSEQ")) {
+                        highestModSeq = sr.readLong();
+                    } else if (responseCode.equalsIgnoreCase("MODIFIED")) {
+                        modifiedMsgsets = MessageNumberSet.buildMessageNumberSets(sr.readAtom());
+                    }
+                }
+            }
+
+            return new StoreResult(highestModSeq, fetchResponses, modifiedMsgsets);
+        }
+
+        /**
+         * Parses the responses from Fetch command and UID Fetch command to a {@link FetchResult} object.
+         *
+         * @param ir the list of responses from Fetch or UID Fetch command, the input responses array should contain the tagged/final one
+         * @param extensionItems the array of extension FetchItem
+         * @return FetchResult object constructed based on the given IMAPResponse array,
+         * @throws ImapAsyncClientException when tagged response is not OK or given response length is 0
+         * @throws ProtocolException when fetch response cannot be parsed
+         */
+        @Nonnull
+        private FetchResult parseToFetchResult(@Nonnull final IMAPResponse[] ir, @Nonnull final FetchItem[] extensionItems)
+                throws ImapAsyncClientException, ProtocolException {
+            if (ir.length < 1) {
+                throw new ImapAsyncClientException(FailureType.INVALID_INPUT);
+            }
+            final Response taggedResponse = ir[ir.length - 1];
+            if (!taggedResponse.isOK()) {
+                throw new ImapAsyncClientException(FailureType.INVALID_INPUT);
+            }
+            final List<FetchResponse> fetchResponses = new ArrayList<>(); // will always return a non-null array
+
+            for (final IMAPResponse sr: ir) {
+                if (sr.keyEquals("FETCH")) {
+                    fetchResponses.add(parseToFetchResponse(sr, extensionItems));
+                }
+            }
+
+            return new FetchResult(fetchResponses);
+        }
+
+        /**
+         * This methods converts a response to {@code FetchResponse}.
+         *
+         * @param response ImapResponse to be converted to Fetch response
+         * @param extensionItems the array of extension FetchItem
+         * @return parsed response if FetchResponse is a fetch response
+         * @throws ProtocolException when encountering error
+         */
+        @Nonnull
+        private FetchResponse parseToFetchResponse(@Nonnull final IMAPResponse response, @Nonnull final FetchItem[] extensionItems)
```

Comment:
remove this method

Reply:
make the code inline, it is easier to know the flow and there is no point to have only one line

Reply:
The main point of helper function is to convert IOException to ProtocolException which will be used for several result objects and future result objects. I can just add throw IOException for each function. However, since readValue needs to add this as well, and it will require all the existing client side projects to make change which is unnecessary.